src/main/java/org/folio/linked/data/mapper/kafka/impl/IndexIdentifierMapperImpl.java

```diff
@@ -0,0 +1,127 @@
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.folio.linked.data.mapper.dto.common.SingleResourceMapper;
+import org.folio.linked.data.mapper.kafka.IndexIdentifierMapper;
+import org.folio.linked.data.model.entity.Resource;
+import org.folio.linked.data.model.entity.ResourceEdge;
+import org.folio.linked.data.model.entity.ResourceTypeEntity;
+
+@Log4j2
+@RequiredArgsConstructor
+public class IndexIdentifierMapperImpl<T, E extends Enum<E>> implements IndexIdentifierMapper<T> {
+
+  private static final String MSG_UNKNOWN_TYPES =
+    "Unknown type(s) [{}] of [{}] was ignored during Resource [resourceId = {}] conversion to BibframeIndex message";
+
+  private static final Collection<String> IDENTIFIER_PROPERTY_VALUES = List.of(
+    NAME.getValue(),
+    EAN_VALUE.getValue(),
+    LOCAL_ID_VALUE.getValue()
+  );
+
+  private final SingleResourceMapper singleResourceMapper;
+  private final Class<?> parentDto;
+  private final Class<? extends Enum<?>> enumClass;
+  private final Function<String, E> typeSupplier;
+  private final Function<String, T> indexCreateByValueFunction;
+  private final BiFunction<T, E, T> indexUpdateTypeFunction;
+
+  @Override
+  public List<T> extractIdentifiers(Resource resource) {
+    return resource.getOutgoingEdges()
+      .stream()
+      .filter(re -> MAP.getUri().equals(re.getPredicate().getUri()))
+      .map(ResourceEdge::getTarget)
+      .map(this::mapToIdentifier)
+      .flatMap(Optional::stream)
+      .distinct()
+      .toList();
+  }
+
+  private Optional<T> mapToIdentifier(Resource resource) {
+    return Optional.of(resource)
+      .map(Resource::getDoc)
+      .flatMap(this::getValue)
+      .map(indexCreateByValueFunction)
+      .map(t -> appendType(resource, t));
+  }
+
+  private Optional<String> getValue(JsonNode doc) {
+    return IDENTIFIER_PROPERTY_VALUES.stream()
+      .filter(doc::has)
+      .map(doc::get)
+      .filter(node -> !node.isEmpty())
+      .map(value -> value.get(0))
+      .map(JsonNode::asText)
+      .findFirst();
+  }
+
+  private T appendType(Resource resource, T t) {
+    return toType(resource)
+      .map(type -> indexUpdateTypeFunction.apply(t, type))
+      .orElse(t);
+  }
+
+  private Optional<E> toType(Resource resource) {
+    if (isNull(resource.getTypes())) {
+      return Optional.empty();
+    }
+    var type = resource.getTypes()
+      .stream()
+      .map(ResourceTypeEntity::getUri)
+      .filter(this::isTypePresent)
+      .findFirst()
+      .map(this::normalizeUri)
+      .flatMap(this::getE);
+
+    if (type.isEmpty()) {
+      logError(resource);
+    }
+    return type;
+  }
+
+  private Optional<E> getE(String typeUri) {
+    try {
+      return Optional.of(typeSupplier.apply(typeUri));
+    } catch (IllegalArgumentException ignored) {
+      return Optional.empty();
+    }
+  }
+
+  private boolean isTypePresent(String typeUri) {
+    return singleResourceMapper.getMapperUnit(typeUri, MAP, parentDto, null)
```

Comment:
I know this is an existing code. However, I forgot why we did this. 
I think following logic is enough in `toType` method.


```
  private Optional<E> toType(Resource resource) {
    if (isNull(resource.getTypes())) {
      return Optional.empty();
    }
    var type = resource.getTypes()
      .stream()
      .filter(t -> t != ResourceTypeDictionary.IDENTIFIER) // Add this line. 
      .map(ResourceTypeEntity::getUri)
      .findFirst()
      .map(this::normalizeUri)
      .flatMap(this::getE);

    if (type.isEmpty()) {
      logError(resource);
    }
    return type;
  }
```

Reply:
I think I remember now.

Earlier, we had data loaded through Python ETL in our graph. Those resources had types that we did not support & hence the code failed. So, we added the logic -- `singleResourceMapper.getMapperUnit(typeUri, MAP, parentDto, null).isPresent()`

Now that we are going to replace Python ETL with Java, I think we can simplify the logic as I gave above. 

What do you think @PBobylev?